s3box_review/s3box.go

```diff
@@ -0,0 +1,243 @@
+package s3box
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	// DefaultBufferSize is set to 10MB
+	DefaultBufferSize = 10000000
+)
+
+var (
+	// ErrS3BucketRequired signals an s3 bucket wasn't provided
+	ErrS3BucketRequired = fmt.Errorf("An s3 bucket is required to create an s3box.")
+
+	// ErrBoxIsSealed signals an operation which can't occur when a box is sealed.
+	ErrBoxIsSealed = fmt.Errorf("Cannot perform action when box is sealed.")
+
+	// ErrBoxNotSealed signals an operation which can't occur when a box is not sealed.
+	ErrBoxNotSealed = fmt.Errorf("Cannot perform action when box is not sealed.")
+
+	// ErrInvalidJSONInput captures when the input data can't be marshalled into JSON.
+	ErrInvalidJSONInput = fmt.Errorf("Only JSON-able inputs are supported for syncing to Redshift.")
+)
+
+// S3Box manages piping data into S3. The mechanics are to buffer data locally, ship to s3 when too much is buffered, and finally create manifests pointing to the data files.
+type S3Box struct {
+	// Inheret mutex locking/unlocking
+	sync.Mutex
+
+	// s3Bucket specifies the intermediary bucket before ultimately piping to Redshift. The user should have access to this bucket.
+	s3Bucket string
+
+	// s3Handler manages the piping of data to s3
+	s3Handler *s3.S3
+
+	// bufferSize is the maximum size of data we're willing to buffer before creating an s3 file
+	bufferSize int
+
+	// bufferedData is the data currently buffered in the box. Calling Dump ships this data into s3
+	bufferedData []byte
+
+	// timestamp tracks the time a box was created or reset
+	timestamp time.Time
+
+	// fileNumber indicates the number of s3 files which have currently been created
+	fileNumber int
```

Comment:
It is, just felt it was nicer not to always have to write `len(fileLocations)`. However there's no issue in removing this redundancy.